stdlib/asm/math/poly512.masm

```diff
@@ -0,0 +1,638 @@
+# - {f, g}_addr0 -> {f, g}_start_addr
+#
+# Output stack state :
+#
+# [h_start_addr, ...]
+#
+# - h_addr`i` is the address of four elements | h[ (i << 2) .. ((i+1) << 2) ) & i ∈ [0..128)
+#
+# Consecutive 127 memory addresses can be computed from starting memory address ( living on stack top ) by 
+# continuing to apply `INCR` ( = add.1 ) instruction on previous absolute memory address.
+#
+# Note, input memory addresses are considered to be read-only, they are not mutated.
+export.mul_zq.128
+    exec.ntt512::forward
+
+    push.env.locaddr.127
+    push.0.0.0.0
+
+    repeat.128
+        dup.5
+        loadw.mem
+
+        dup.4
+        storew.mem
+
+        movup.5
+        add.1
+        movdn.5
+
+        movup.4
+        add.1
+        movdn.4
+    end
+
+    dropw
+    drop
+    drop
+
+    exec.ntt512::forward
+
+    push.env.locaddr.127
+    push.0.0.0.0.0.0.0.0
+
+    repeat.128
+        dup.9
+        loadw.mem
+
+        swapw
+
+        dup.8
+        loadw.mem
+
+        exec.mul_word
+
+        dup.4
+        storew.mem
+
+        movup.5
+        add.1
+        movdn.5
+
+        movup.4
+        add.1
+        movdn.4
+
+        push.0.0.0.0
+    end
+
+    dropw
+    dropw
+    drop
+    drop
+
+    push.env.locaddr.127
+
+    exec.ntt512::backward
+
+    push.env.locaddr.127
+    push.0.0.0.0
+
+    repeat.128
+        dup.5
+        loadw.mem
+
+        exec.reduce_word
+
+        dup.4
+        storew.mem
+
+        movup.5
+        add.1
+        movdn.5
+
+        movup.4
+        add.1
+        movdn.4
+    end
+
+    dropw
+    drop
+    drop
+
+    push.env.locaddr.127
+end
+
+# Given two polynomials of degree 512 on stack as absolute memory addresses,
+# this routine computes polynomial addition.
+#
+# Imagine, two polynomials f, g
+#
+# h = f + g, can be computed as
+#
+# [(f[i] + g[i]) % Q for i in range(512)] | Q = 12289 ( = Falcon Digital Signature Algorithm's Prime Number )
+#
+# Input stack state :
+#
+# [f_start_addr, g_start_addr, ...]
+#
+# - {f, g}_addr`i` -> {f, g}[ (i << 2) .. ((i+1) << 2) ), address holding four consecutive coefficients
+# - {f, g}_addr0 -> {f, g}_start_addr
+#
+# Output stack state :
+#
+# [h_start_addr, ...]
+#
+# - h_addr`i` is the address of four elements | h[ (i << 2) .. ((i+1) << 2) ) & i ∈ [0..128)
+#
+# Consecutive 127 memory addresses can be computed from starting memory address ( living on stack top ) by 
+# continuing to apply `INCR` ( = add.1 ) instruction on previous absolute memory address.
+#
+# Note, input memory addresses are considered to be read-only, they are not mutated.
+export.add_zq.128
+    push.env.locaddr.127
+    push.0.0.0.0.0.0.0.0
+
+    repeat.128
+        dup.9
+        loadw.mem
+
+        swapw
+
+        dup.10
+        loadw.mem
+
+        exec.add_word
+        exec.mod_div_word
+
+        dup.4
+        storew.mem
+
+        movup.6
+        add.1
+        movdn.6
```

Comment:
I wonder if we could change this to load the coefficients for g polynomial immediately. Maybe something like this:
```
movup.6
add.1
dup
movdn.6
loadw.mem
```
If this can let us avoid using `dup.10` above this could save quite a few cycles (`dup.10` actually takes 3 cycles, while `dup.9` only takes 1 cycle).

Reply:
Tried rewriting, but looks like both `dup.9` & `dup.10` costs 3 cycles.